MaaSSim/d2d.py

```diff
@@ -265,13 +323,46 @@ def pax_mode_choice(*args, **kwargs):
     # returns boolean True if passenger decides not to use (private) ridesourcing for a given day (i.e. low quality offer)
     traveller = kwargs.get('traveller',None)
     sim = traveller.sim
+    params = sim.params
+    mcp = params.mode_choice
+    mset = params.alt_modes
+    mode_choice = kwargs.get('mode_choice',True)
     
     platform_id, offer = list(traveller.offers.items())[0]
-    delta = 0.5
     
-    pass_walk_time = sim.skims.walk[traveller.pax.pos][traveller.request.origin]
-    veh_pickup_time = sim.skims.ride.T[sim.vehs[offer['veh_id']].veh.pos][traveller.request.origin]
-    pass_matching_time = sim.env.now - traveller.t_matching
-    tt = traveller.request.ttrav
+    if mode_choice == True:
+        
+        rs_wait = sim.skims.ride.T[sim.vehs[offer['veh_id']].veh.pos][traveller.request.origin]
+        rs_ivt = traveller.request.ttrav.seconds
+        rs_fare = max(params.platforms.base_fare + params.platforms.fare * rs_ivt * (params.speeds.ride / 1000), params.platforms.min_fare)
+        
+        car_ivt = traveller.request.ttrav.seconds # assumed same as RS
+        car_cost = mset.car.km_cost * car_ivt * (params.speeds.ride / 1000)
+        pt_ivt = traveller.request.ttrav.seconds # assumed same as RS
+        pt_fare = mset.pt.base_fare + mset.pt.km_fare * pt_ivt * (params.speeds.ride / 1000)
+        
+        bike_tt = sim.skims.ride.T[traveller.request.origin][traveller.request.destination] * (params.speeds.ride / params.speeds.bike)
+        
+        U_rs = mcp.beta_wait_rs * rs_wait + mcp.beta_time_moto * rs_ivt + mcp.beta_cost * rs_fare + mcp.ASC_rs
+        U_car = mcp.beta_access * mset.car.access_time + mcp.beta_time_moto * car_ivt + mcp.beta_cost * car_cost + mcp.ASC_car
+        U_pt = mcp.beta_access * mset.pt.access_time + mcp.beta_wait_pt * mset.pt.wait_time + mcp.beta_time_moto * pt_ivt + mcp.beta_cost * pt_fare + mcp.ASC_pt
+        U_bike = mcp.beta_time_bike * bike_tt
+        U_list = [U_rs, U_car, U_pt, U_bike]
+        
+        P_list = [np.exp(U_list[mode_id]) / (np.exp(U_rs) + np.exp(U_car) + np.exp(U_pt) + np.exp(U_bike)) for mode_id in range(len(U_list))]
```

Comment:
this sum may be computed once, before the loop of modes. it will be faster

Reply:
The denominator you mean?